sqlstats: move meta fields out of stmtKey

xinhaoz opened the pull request:
Previously the following 3 fields were part of the
key in the stmtStats map:
- query fingerprint (string)
- database (string)
- implicit txn (bool)

To slightly reduce the overhead of looking into the map,
an operation we perform frequently, we can replace all these
fields with the stmt fingerprint id (uint64).

This commit also does the following:
- Updates TestSQLStatsPlanSampling - updating that
to TestSQLStatsStmtSampling and making it more accurately
reflect the goal of the underlying cache it is testing.

```
name                               old time/op    new time/op    delta
StmtKeyMapOperations/MapInsert-10    1.38ms ± 2%    0.67ms ± 3%  -51.41%  (p=0.000 n=10+10)
StmtKeyMapOperations/MapLookup-10     544µs ± 1%     267µs ± 2%  -50.83%  (p=0.000 n=10+9)

name                               old alloc/op   new alloc/op   delta
StmtKeyMapOperations/MapInsert-10    2.48MB ± 0%    1.29MB ± 0%  -48.07%  (p=0.000 n=10+9)
StmtKeyMapOperations/MapLookup-10     0.00B          0.00B          ~     (all equal)

name                               old allocs/op  new allocs/op  delta
StmtKeyMapOperations/MapLookup-10      0.00           0.00          ~     (all equal)
StmtKeyMapOperations/MapInsert-10       266 ± 0%       274 ± 0%   +3.01%  (p=0.000 n=10+10)
```

Epic: none

Release note: None

xinhaoz committed:
sqlstats: move meta fields out of stmtKey

Previously the following 3 fields were part of the
key in the stmtStats map:
- query fingerprint (string)
- database (string)
- implicit txn (bool)

To slightly reduce the overhead of looking into the map,
an operation we perform frequently, we can replace all these
fields with the stmt fingerprint id (uint64).

This commit also does the following:
- Updates TestSQLStatsPlanSampling - updating that
to TestSQLStatsStmtSampling and making it more accurately
reflect the goal of the underlying cache it is testing.

```
name                               old time/op    new time/op    delta
StmtKeyMapOperations/MapInsert-10    1.38ms ± 2%    0.67ms ± 3%  -51.41%  (p=0.000 n=10+10)
StmtKeyMapOperations/MapLookup-10     544µs ± 1%     267µs ± 2%  -50.83%  (p=0.000 n=10+9)

name                               old alloc/op   new alloc/op   delta
StmtKeyMapOperations/MapInsert-10    2.48MB ± 0%    1.29MB ± 0%  -48.07%  (p=0.000 n=10+9)
StmtKeyMapOperations/MapLookup-10     0.00B          0.00B          ~     (all equal)

name                               old allocs/op  new allocs/op  delta
StmtKeyMapOperations/MapLookup-10      0.00           0.00          ~     (all equal)
StmtKeyMapOperations/MapInsert-10       266 ± 0%       274 ± 0%   +3.01%  (p=0.000 n=10+10)
```

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141445)
<!-- Reviewable:end -->

dhartunian reviewed (commented):
> Reviewed 1 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141445)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go` line 456 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OJUYX2U3JnzwLXM9X5I:-OJUYX2U3JnzwLXM9X5J:b6qvzp9) ([raw file](https://github.com/cockroachdb/cockroach/blob/945d97ebf2154c32b5d28b0654af197d98c12a1c/pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go#L456)):*
> ```Go
> 	stats.ID = stmtFingerprintID
> 	s.mu.stmts[key] = stats
> 	s.mu.sampledStatementCache[key.sampledPlanKey] = struct{}{}
> ```

do we need to preserve this cache insert?
___
*[`pkg/sql/sqlstats/persistedsqlstats/flush_test.go` line 820 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OJUZfEz5BD3ASiOuedz:-OJUZfEz5BD3ASiOuee-:b-ygxja7) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6cc443537aedc25ea5c13427c7eb8f5283c63e5/pkg/sql/sqlstats/persistedsqlstats/flush_test.go#L820)):*
> ```Go
> 
> 		checkSampled("SELECT _", true /* implicit */, true /* sampled */)
> 		checkSampled("SELECT _, _", true /* implicit */, true /* sampled */)
> ```

should the `implicit` arg be `false` in the two checks above?
___
*[`pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go` line 638 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OJUYq_B6slRer66uU0w:-OJUYq_B6slRer66uU0x:b-pqu51h) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6cc443537aedc25ea5c13427c7eb8f5283c63e5/pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go#L638)):*
> ```Go
> 			statCopy.mu.distSQLUsed = v.mu.distSQLUsed
> 			statCopy.mu.vectorized = v.mu.vectorized
> 			statCopy.mu.querySummary = v.mu.querySummary
> ```

how would these get set before?


<!-- Sent from Reviewable.io -->

xinhaoz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141445)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

___
*[`pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go` line 456 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OJUYX2U3JnzwLXM9X5I:-OJUuwoi9hOBx14l-Nui:bx61ewx) ([raw file](https://github.com/cockroachdb/cockroach/blob/945d97ebf2154c32b5d28b0654af197d98c12a1c/pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go#L456)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

do we need to preserve this cache insert?
</blockquote></details>

Nope - we insert into the cache at the time of sampling decision in instrumentationHelper.Setup.
___
*[`pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go` line 638 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OJUYq_B6slRer66uU0w:-OJUv77B2AP632wFDIxW:b-3d51l8) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6cc443537aedc25ea5c13427c7eb8f5283c63e5/pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go#L638)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

how would these get set before?
</blockquote></details>

It seemed to me like they weren't getting set before... we only use Add for the reported stats sink so looks like there could have been a bug here and those fields would just been using default falsey values. There's a lack of testing for this function - I can add to verify.
___
*[`pkg/sql/sqlstats/persistedsqlstats/flush_test.go` line 820 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OJUZfEz5BD3ASiOuedz:-OJUwmyOBzXprs5E5_nI:b-xmxv2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6cc443537aedc25ea5c13427c7eb8f5283c63e5/pkg/sql/sqlstats/persistedsqlstats/flush_test.go#L820)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

should the `implicit` arg be `false` in the two checks above?
</blockquote></details>

Oh, yup - nice catch, I'll fix!


<!-- Sent from Reviewable.io -->

xinhaoz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141445)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

___
*[`pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go` line 638 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OJUYq_B6slRer66uU0w:-OJV7gSjABcjSkKm7ixv:b-4769zg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6cc443537aedc25ea5c13427c7eb8f5283c63e5/pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go#L638)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

It seemed to me like they weren't getting set before... we only use Add for the reported stats sink so looks like there could have been a bug here and those fields would just been using default falsey values. There's a lack of testing for this function - I can add to verify.
</blockquote></details>

Ah okay it seems like this function only cares about adding `mu.data` field from each statement stats into the receiver container, so it didn't matter that these fields weren't set before. I can add a quick test still since it doesn't seem like this existed before.


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> Reviewed 1 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141445)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go` line 456 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OJUYX2U3JnzwLXM9X5I:-OJdQSwLEdGnqrYfvQe8:b-9w81lg) ([raw file](https://github.com/cockroachdb/cockroach/blob/945d97ebf2154c32b5d28b0654af197d98c12a1c/pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go#L456)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Nope - we insert into the cache at the time of sampling decision in instrumentationHelper.Setup.
</blockquote></details>

So this was just redundant?
___
*[`-- commits` line 10 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OJ_9bSm0T87WNfmnjyX:-OJ_9bSm0T87WNfmnjyY:b-gy1d9u):*
If your goal is map lookup overhead, can you post a benchmark that illustrates the gains?
___
*[`pkg/sql/sqlstats/ssmemstorage/ss_mem_writer_test.go` line 203 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OJdSn2w9WET1y9h5byT:-OJdSn2w9WET1y9h5byU:b2y4ti8) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e11698733a7f75e00dca1f40965fcbb6eb453/pkg/sql/sqlstats/ssmemstorage/ss_mem_writer_test.go#L203)):*
> ```Go
> 	require.Equal(t, destStmtStats.mu.data.Count, int64(count))
> 	require.Equal(t, destStmtStats.mu.data.FailureCount, int64(count))
> 	require.Less(t, math.Abs(float64(stmtStats.RowsAffected)-destStmtStats.mu.data.NumRows.Mean), epsilon)
> ```

you can use `require.InEpsilon` which will read more easily
___
*[`pkg/sql/sqlstats/ssmemstorage/ss_mem_writer.go` line 135 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OJdRttJC8E-FADMGFHV:-OJdRttJC8E-FADMGFHW:b-dzg8e) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e11698733a7f75e00dca1f40965fcbb6eb453/pkg/sql/sqlstats/ssmemstorage/ss_mem_writer.go#L135)):*
> ```Go
> 		estimatedMemoryAllocBytes += timestampSize + 8
> 		s.mu.Lock()
> 		defer s.mu.Unlock()
> ```

This lock is already taken below in the conditional, should that be removed?
___
*[`pkg/sql/logictest/testdata/logic_test/statement_statistics` line 237 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OJdOsEM0Nj--g3dtgsa:-OJdOi229j9z23EyrYKB:bckld6d) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e11698733a7f75e00dca1f40965fcbb6eb453/pkg/sql/logictest/testdata/logic_test/statement_statistics#L237)):*
> ```
> 10408858296841757964  4544145443417805523   SELECT x FROM test WHERE y = _  true
> 10543509969653713415  646847662919799570    SELECT _                        false
> 10543509969653713419  646847662919799570    SELECT _                        false
> ```

why does `SELECT _` appear twice now? it looks like txn is the same ID and fingerprint is the same.
___
*[`pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go` line 52 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OJ_9nvCE1bToqJH7YV-:-OJ_9nvCE1bToqJH7YV0:b-h79bmm) ([raw file](https://github.com/cockroachdb/cockroach/blob/b30be14916ea6a0a87df0de71fa0c429e10673fb/pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go#L52)):*
> ```Go
> }
> 
> func (s stmtKey) String() string {
> ```

are you sure this is unused? maybe used for logging? 


<!-- Sent from Reviewable.io -->

xinhaoz commented:
*[`pkg/sql/logictest/testdata/logic_test/statement_statistics` line 237 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OJdOsEM0Nj--g3dtgsa:-OJdm0LB09R-7Ahnb7kM:b9lkkk7) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e11698733a7f75e00dca1f40965fcbb6eb453/pkg/sql/logictest/testdata/logic_test/statement_statistics#L237)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

why does `SELECT _` appear twice now? it looks like txn is the same ID and fingerprint is the same.
</blockquote></details>

Ah it looks like before this is due to the upgrade from implicit -> explicit txn in the `SELECT 1; BEGIN; ...` stmt above.
Since the stmt starts off as Implicit, it calculates the stmtFingerprintID incorrectly since it uses implicit=true.
Before this change, while we were still returning the incorrect fingerprintID from RecordStatement, the merge step into the transaction container would still be correct since it uses the key struct and not the id.
Before:
```
stmtBufForCurrentTxn: [ {stmtKey: {implicit: true, query: "SELECT 1"} ]
// BEGIN stmt - upgrade to explicit txn
// At merge step, we set implicit based on the end txn state, so the stmtKey is properly set again.
stmtBufForCurrentTxn: [ {stmtKey: {implicit: false, query: "SELECT 1"} ]
```

Now:
```
stmtBufForCurrentTxn: [ {stmtKey: <generatedIDWithEncodedImplicit> ]
// BEGIN stmt - upgrade to explicit txn, but we don't go back and correct stmts already executed.
stmtBufForCurrentTxn: [ {stmtKey: <generatedIDWithEncodedImplicit> ]
```

In the short term, I can patch this PR that we go back and recalculate the fingerprintid on upgrade to explicit txn - that scenario shouldn't often. We were doing something similar already but I got rid of it when switching to use a slice as the stmt buf over the ssmemstorage.Container.
Personally I'd actually like to get rid of `implicit` as part of the statement fingerprint id and as a property of the collected stats (this would be a separate PR). I feel like tracking that property and bucketing does not add much value. WDYT?


<!-- Sent from Reviewable.io -->

xinhaoz commented:
Very small memory improvement, could just be noise though. I'll add to the commit msg as well:
```
old:  fcfb62d Merge #142185
new:  7e5ca5b sqlstats: move meta fields out of stmtKey
args: benchdiff "./pkg/sql/tests" "-b" "-r" "Sysbench/SQL/3node/oltp_read_write" "-d" "1000x" "-c" "15" "--cpuprofile" "--memprofile"

name                                           old time/op    new time/op    delta
Sysbench/SQL/3node/oltp_read_write-24            12.7ms ± 1%    12.8ms ± 3%    ~     (p=1.000 n=15+15)
ParallelSysbench/SQL/3node/oltp_read_write-24    1.20ms ± 6%    1.22ms ± 6%    ~     (p=0.186 n=14+15)

name                                           old errs/op    new errs/op    delta
Sysbench/SQL/3node/oltp_read_write-24              0.00           0.00         ~     (all equal)
ParallelSysbench/SQL/3node/oltp_read_write-24      0.01 ±30%      0.01 ±57%    ~     (p=0.922 n=14+15)

name                                           old alloc/op   new alloc/op   delta
ParallelSysbench/SQL/3node/oltp_read_write-24    2.04MB ± 1%    2.03MB ± 1%  -0.73%  (p=0.004 n=12+15)
Sysbench/SQL/3node/oltp_read_write-24            2.19MB ± 1%    2.18MB ± 1%  -0.41%  (p=0.014 n=13+12)

name                                           old allocs/op  new allocs/op  delta
Sysbench/SQL/3node/oltp_read_write-24             10.6k ± 1%     10.6k ± 1%    ~     (p=0.305 n=14+15)
ParallelSysbench/SQL/3node/oltp_read_write-24     9.27k ± 2%     9.25k ± 2%    ~     (p=0.382 n=14+14)
```

xinhaoz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141445)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

___
*[`pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go` line 456 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OJUYX2U3JnzwLXM9X5I:-OJdyQ0B5iog2NMA3JtD:biuf7zi) ([raw file](https://github.com/cockroachdb/cockroach/blob/945d97ebf2154c32b5d28b0654af197d98c12a1c/pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go#L456)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

So this was just redundant?
</blockquote></details>

It became redundant since https://github.com/cockroachdb/cockroach/pull/140399.
___
*[`pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go` line 52 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OJ_9nvCE1bToqJH7YV-:-OJdlZML3IL6po9ExHot:bdpqu6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/b30be14916ea6a0a87df0de71fa0c429e10673fb/pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go#L52)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

are you sure this is unused? maybe used for logging?
</blockquote></details>

Yeah I searched for all use cases and it turned up empty.
___
*[`pkg/sql/sqlstats/ssmemstorage/ss_mem_writer.go` line 135 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OJdRttJC8E-FADMGFHV:-OJdluXQA2fGlRstzE7m:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e11698733a7f75e00dca1f40965fcbb6eb453/pkg/sql/sqlstats/ssmemstorage/ss_mem_writer.go#L135)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

This lock is already taken below in the conditional, should that be removed?
</blockquote></details>

Done.
___
*[`pkg/sql/sqlstats/ssmemstorage/ss_mem_writer_test.go` line 203 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OJdSn2w9WET1y9h5byT:-OJdx67M8HbOZJ_bJ7m4:b-efu71k) ([raw file](https://github.com/cockroachdb/cockroach/blob/208e11698733a7f75e00dca1f40965fcbb6eb453/pkg/sql/sqlstats/ssmemstorage/ss_mem_writer_test.go#L203)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

you can use `require.InEpsilon` which will read more easily
</blockquote></details>

Oh good shout, thanks! Done.


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> Reviewed 4 of 11 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141445)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`-- commits` line 14 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OLKgHDc2Yz_qmCiNMFy:-OLKgHDc2Yz_qmCiNMFz:b-a38yvi):*
Something for next time, but if you find yourself writing "this commit also...", that's perhaps a signal that you should have a separate commit. Even in a single PR splitting stuff up makes the review easier. Right now I have to look at a mixed diff of all this stuff and need to separate the streams mentally.
___
*[`-- commits` line 17 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OLKeLJhELQkkysAamjE:-OLKeLJi0znN5xHgsTM4:b-2c9t88):*
In the future, I would suggest not mixing two changes like this together because if you're measuring perf impact there's no way to know which part of your diff made the contribution. Below we see an alloc difference, and I have no idea whether it's to do with the map key or the memory pool.
___
*[`pkg/sql/sqlstats/sslocal/sslocal_stats_collector.go` line 185 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OLKvwtp6EXd-kZkRBoV:-OLKvwtp6EXd-kZkRBoW:b-dq9f1q) ([raw file](https://github.com/cockroachdb/cockroach/blob/a53584b61caf2f37d2de9cfa46ac80cae4103f75/pkg/sql/sqlstats/sslocal/sslocal_stats_collector.go#L185)):*
> ```Go
> 	for _, stmt := range s.stmtBuf {
> 		stmt.TransactionFingerprintID = transactionFingerprintID
> 		stmt.ImplicitTxn = implicitTxn
> ```

Here we set `ImplicitTxn` from the args, don't we need to recompute the fingerprintID as required by `UpgradeToExplicitTransaction`. 

Additionally, if the flag is set in `EndTransaction` and we iterate through all of `stmtBuf` anyway, why do we even need `UpgradeToExplicitTransaction` at all? I assume in these cases `writeDirectlyToFlushTarget` is false and the `RecordedStmtStats` structs stick around for modification.
___
*[`pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go` line 216 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OLL07nh3VldCikx95Pu:-OLL07nh3VldCikx95Pv:b-a7g96q) ([raw file](https://github.com/cockroachdb/cockroach/blob/a53584b61caf2f37d2de9cfa46ac80cae4103f75/pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go#L216)):*
> ```Go
> 				implicitTxn:     statistics[i].Key.KeyData.ImplicitTxn,
> 				database:        statistics[i].Key.KeyData.Database,
> 			})
> ```

It looks like passing in `sampledPlanKey` here sets it on `stmtStats` within `tryCreateStatsForStmtWithKeyLocked`, yet below we set the same fields again. Why? Am I missing something?
___
*[`pkg/sql/conn_executor.go` line 4050 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OLKgLUh4FzV6mD7UwQl:-OLKgLUh4FzV6mD7UwQm:bfvznoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/a53584b61caf2f37d2de9cfa46ac80cae4103f75/pkg/sql/conn_executor.go#L4050)):*
> ```Go
> 			if advInfo.txnEvent.eventType == txnUpgradeToExplicit {
> 				ex.extraTxnState.txnFinishClosure.implicit = false
> 				ex.statsCollector.UpgradeToExplicitTransaction()
> ```

This kind of code worries me. We need to magically "know" to call `UpgradeToExplicitTransaction()` for things to work correctly. We should try and control the "sprawl" of `statsCollector` within `conn_executor.go`. I have a separate comment later that questions the need for this.


<!-- Sent from Reviewable.io -->

xinhaoz commented:
Also, here's the benchmark of the map insert itself (can include the code for that in the PR if desired). It's only a small improvement but I think it'll also open the door for us to more improvements such as introducing a stmtKey -> fingerprintID cache, which fits into the stmt meta normalization work we want to pursue.

args: benchdiff "./pkg/sql/sqlstats/ssmemstorage" "-b" "-r" "MapOperations" "-d" "10000x" "-c" "10" "--cpuprofile" "--memprofile"

name                        old time/op    new time/op    delta
MapOperations/MapLookup-10     557µs ± 2%     265µs ± 2%  -52.42%  (p=0.000 n=10+9)
MapOperations/MapInsert-10    1.36ms ±10%    0.65ms ± 3%  -52.21%  (p=0.000 n=10+10)

name                        old alloc/op   new alloc/op   delta
MapOperations/MapLookup-10    1.30B ±131%    0.50B ±100%  -61.54%  (p=0.039 n=10+10)
MapOperations/MapInsert-10    2.48MB ± 0%    1.29MB ± 0%  -48.07%  (p=0.000 n=10+10)

name                        old allocs/op  new allocs/op  delta
MapOperations/MapLookup-10      0.00           0.00          ~     (all equal)
MapOperations/MapInsert-10       266 ± 0%       274 ± 0%   +3.01%  (p=0.000 n=10+10)

xinhaoz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141445)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

___
*[`-- commits` line 17 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OLKeLJhELQkkysAamjE:-OLLTjv6CxlTSHYOCyn0:by41vg2):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

In the future, I would suggest not mixing two changes like this together because if you're measuring perf impact there's no way to know which part of your diff made the contribution. Below we see an alloc difference, and I have no idea whether it's to do with the map key or the memory pool.
</blockquote></details>

Ack - I split those changes out from this PR.
___
*[`pkg/sql/conn_executor.go` line 4050 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OLKgLUh4FzV6mD7UwQl:-OLLRw6Q41bwuZnhHUAO:b-8bszkg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a53584b61caf2f37d2de9cfa46ac80cae4103f75/pkg/sql/conn_executor.go#L4050)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

This kind of code worries me. We need to magically "know" to call `UpgradeToExplicitTransaction()` for things to work correctly. We should try and control the "sprawl" of `statsCollector` within `conn_executor.go`. I have a separate comment later that questions the need for this.
</blockquote></details>

Yeah I don't disagree - this was the mitigation we had earlier to deal with transactions being upgraded to explicit. I think when we move everything to async, we can get rid of this.
___
*[`pkg/sql/sqlstats/sslocal/sslocal_stats_collector.go` line 185 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OLKvwtp6EXd-kZkRBoV:-OLLSdg0BpeDXUukpCVY:b-qof494) ([raw file](https://github.com/cockroachdb/cockroach/blob/a53584b61caf2f37d2de9cfa46ac80cae4103f75/pkg/sql/sqlstats/sslocal/sslocal_stats_collector.go#L185)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Here we set `ImplicitTxn` from the args, don't we need to recompute the fingerprintID as required by `UpgradeToExplicitTransaction`. 

Additionally, if the flag is set in `EndTransaction` and we iterate through all of `stmtBuf` anyway, why do we even need `UpgradeToExplicitTransaction` at all? I assume in these cases `writeDirectlyToFlushTarget` is false and the `RecordedStmtStats` structs stick around for modification.
</blockquote></details>

The fingerprint ID calculation is currently handled in `connExecutor.recordStatementSummary` because it's also sent to insights collection. There is an existing bug here: when a transaction is upgraded from implicit to explicit, the fingerprint ID sent to insights will differ (and is incorrect) from what's recorded in SQL stats, where we recompute the correct value on upgrade.

The `UpgradeToExplicitTransaction` function exists so that we we only recalculate when necessary, instead of once for insights and once at EndTransaction.

I agree that this isn't ideal and creates some intermediate complexit. In the next PRs we'll unify stats and insights recording into the same asynchronous process and move the fingerprint id calculation there to avoid this.

We could simplify now if we accept some incorrect fingerprint IDs in SQL stats for upgraded transactions.

For the current PR, I've moved the setting of the implicit flag to the upgrade call, which addresses the immediate concern while maintaining the existing behavior.
___
*[`pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go` line 216 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OLL07nh3VldCikx95Pu:-OLLVbNs8qEDqBZxTFhy:b-si9ybb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a53584b61caf2f37d2de9cfa46ac80cae4103f75/pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go#L216)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

It looks like passing in `sampledPlanKey` here sets it on `stmtStats` within `tryCreateStatsForStmtWithKeyLocked`, yet below we set the same fields again. Why? Am I missing something?
</blockquote></details>

No you're right we don't need both - removed.


<!-- Sent from Reviewable.io -->

xinhaoz committed:
ssmemstorage: add microbenchmark for stmtKey

Add `BenchmarkStmtKeyMapOperations` for a map with
`stmtKey` as the key.

Epic: none

Release note: None

dhartunian reviewed (approved):
> Reviewed 3 of 14 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141445)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @xinhaoz)

___
*[`pkg/sql/sqlstats/sslocal/sslocal_stats_collector.go` line 185 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OLKvwtp6EXd-kZkRBoV:-OLZuvir4J6-9soRswID:bo58ywg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a53584b61caf2f37d2de9cfa46ac80cae4103f75/pkg/sql/sqlstats/sslocal/sslocal_stats_collector.go#L185)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

The fingerprint ID calculation is currently handled in `connExecutor.recordStatementSummary` because it's also sent to insights collection. There is an existing bug here: when a transaction is upgraded from implicit to explicit, the fingerprint ID sent to insights will differ (and is incorrect) from what's recorded in SQL stats, where we recompute the correct value on upgrade.

The `UpgradeToExplicitTransaction` function exists so that we we only recalculate when necessary, instead of once for insights and once at EndTransaction.

I agree that this isn't ideal and creates some intermediate complexit. In the next PRs we'll unify stats and insights recording into the same asynchronous process and move the fingerprint id calculation there to avoid this.

We could simplify now if we accept some incorrect fingerprint IDs in SQL stats for upgraded transactions.

For the current PR, I've moved the setting of the implicit flag to the upgrade call, which addresses the immediate concern while maintaining the existing behavior.
</blockquote></details>

Can you write down this explanation somewhere? Maybe in the docsting to `UpgradeToExplicitTransaction`? This information should be in the codebase.
___
*[`pkg/sql/sqlstats/ssmemstorage/ss_mem_writer_test.go` line 240 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OLZu53yC-RJaT5q8Ajp:-OLZu53z73muC3Raw85e:bl0bczk) ([raw file](https://github.com/cockroachdb/cockroach/blob/369af119c3171aa6649b49a06c32bb18f8c3d9c2/pkg/sql/sqlstats/ssmemstorage/ss_mem_writer_test.go#L240)):*
> ```Go
> 			m[keys[j]] = j
> 		}
> 		b.ResetTimer()
> ```

Why do we need to call `ResetTimer` twice?


<!-- Sent from Reviewable.io -->

xinhaoz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141445)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

___
*[`pkg/sql/sqlstats/sslocal/sslocal_stats_collector.go` line 185 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OLKvwtp6EXd-kZkRBoV:-OL_G3J06zDYrq0_IrHi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a53584b61caf2f37d2de9cfa46ac80cae4103f75/pkg/sql/sqlstats/sslocal/sslocal_stats_collector.go#L185)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Can you write down this explanation somewhere? Maybe in the docsting to `UpgradeToExplicitTransaction`? This information should be in the codebase.
</blockquote></details>

Done.
___
*[`pkg/sql/sqlstats/ssmemstorage/ss_mem_writer_test.go` line 240 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/141445#-OLZu53yC-RJaT5q8Ajp:-OL_G4bYB7-2S56ZgkUh:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/369af119c3171aa6649b49a06c32bb18f8c3d9c2/pkg/sql/sqlstats/ssmemstorage/ss_mem_writer_test.go#L240)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Why do we need to call `ResetTimer` twice?
</blockquote></details>

Removed.


<!-- Sent from Reviewable.io -->

xinhaoz commented:
@dhartunian Thanks for the review + patience on this one!

bors r+
